pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/Java15PreviewTest.java

```diff
@@ -0,0 +1,210 @@
+            Assert.assertTrue(variable.isFinal());
+            // Note: these variables are not part of the symbol table
+            // See ScopeAndDeclarationFinder#visit(ASTVariableDeclaratorId, Object)
+            Assert.assertNull(variable.getNameDeclaration());
+        }
+    }
+
+    @Test(expected = ParseException.class)
+    public void patternMatchingInstanceofBeforeJava15PreviewShouldFail() {
+        java15.parseResource("PatternMatchingInstanceof.java");
+    }
+
+    @Test
+    public void recordPoint() {
+        ASTCompilationUnit compilationUnit = java15p.parseResource("Point.java");
+        ASTRecordDeclaration recordDecl = compilationUnit.getFirstDescendantOfType(ASTRecordDeclaration.class);
+        Assert.assertEquals("Point", recordDecl.getImage());
+        Assert.assertFalse(recordDecl.isNested());
+        Assert.assertFalse(recordDecl.isLocal());
+        Assert.assertTrue("Records are implicitly always final", recordDecl.isFinal());
+        List<ASTRecordComponent> components = recordDecl.getFirstChildOfType(ASTRecordComponentList.class)
+                                                        .findChildrenOfType(ASTRecordComponent.class);
+        Assert.assertEquals(2, components.size());
+        Assert.assertEquals("x", components.get(0).getVarId().getImage());
+        Assert.assertEquals("y", components.get(1).getVarId().getImage());
+        Assert.assertNull(components.get(0).getVarId().getNameDeclaration().getAccessNodeParent());
+        Assert.assertEquals(Integer.TYPE, components.get(0).getVarId().getNameDeclaration().getType());
+        Assert.assertEquals("int", components.get(0).getVarId().getNameDeclaration().getTypeImage());
+    }
+
+    @Test(expected = ParseException.class)
+    public void recordPointBeforeJava15PreviewShouldFail() {
+        java15.parseResource("Point.java");
+    }
+
+    @Test(expected = ParseException.class)
+    public void recordCtorWithThrowsShouldFail() {
+        java15p.parse("  record R {"
+                          + "   R throws IOException {}"
+                          + "  }");
+    }
+
+    @Test(expected = ParseException.class)
+    public void recordMustNotExtend() {
+        java15p.parse("record RecordEx(int x) extends Number { }");
+    }
+
+    @Test(expected = ParseException.class)
+    @Ignore("Should we check this?")
+    public void recordCannotBeAbstract() {
+        java15p.parse("abstract record RecordEx(int x) { }");
+    }
+
+    @Test(expected = ParseException.class)
+    @Ignore("Should we check this?")
+    public void recordCannotHaveInstanceFields() {
+        java15p.parse("record RecordFields(int x) { private int y = 1; }");
+    }
+
+    @Test
+    public void innerRecords() {
+        ASTCompilationUnit compilationUnit = java15p.parseResource("Records.java");
+        List<ASTRecordDeclaration> recordDecls = compilationUnit.findDescendantsOfType(ASTRecordDeclaration.class, true);
```

Comment:
I'll try tomorrow to write Kotlin - that basically entails that we ditch eclipse as a supported IDE for developing PMD (see https://github.com/pmd/pmd/wiki/Setup-IDE)... since Kotlin+Eclipse doesn't work.

Reply:
There's also the possibility of writing tests like #2445, which looks useful since your test data is already in separate files. This doesn't need to be in kotlin